scipy/stats/morestats.py

```diff
@@ -2105,9 +2120,34 @@ def ansari(x, y):
 
     Notes
     -----
-    The p-value given is exact when the sample sizes are both less than
-    55 and there are no ties, otherwise a normal approximation for the
-    p-value is used.
+    The ``mode`` parameter is used in the following manner:
+
+    - For ``mode='auto'``, the p-value given is exact the sample sizes
+      are both less than 55 and there are no ties, otherwise a normal
+      approximation for the p-value is used.
+    - For ``mode='exact'``, the p-value given is exact when there are
+      no ties, otherwise a normal approximation for the p-value is used.
```

Comment:
Is the normal approximation any more correct when there are ties? If not, should `exact` always use the exact distribution? We should just warn users that it is exact under the assumption that there are no ties.

Reply:
> Is the normal approximation any more correct when there are ties?

I have not checked but the exact distribution may take a longer time to generate when ties are present? ([I saw this for the wilcoxon signed rank test](https://stats.stackexchange.com/questions/407170/how-does-normal-approximation-work-when-ties-exist-for-wilcoxon-signed-rank-test)) Also, R always falls back to the approximation method when ties are present. Not saying that we should blindly inherit that behavior though. I am OK with both:
  - always using the exact distribution (irrespective of ties) when `mode='exact'`
  - just give a warning and use the exact distribution.

What do you think will be better to do?

Also, for now, I am keeping the part where I throw an error when ties are present and `mode='exact'`. We need to change that once we decide which option is better. I will also update the documentation and tests accordingly.

Reply:
> I have not checked but the exact distribution may take a longer time to generate when ties are present?

I think the post from SO means that if ties are considered in the exact calculation,  calculating the exact distribution takes longer. Our implementation of the exact distribution doesn't seem to account for ties, so my understanding is that execution time would not be affected. Would you agree?

I think what we should do here depends on whether we do some sort of tie correction in the asymptotic method. For instance, our `wilcoxon` does what you have here, but I see that the asymptotic method adjusts for ties (see `n_zero`). From that perspective, I can see why, in the presence of ties, the asymptotic method might be more correct than the exact method. Does the asymptotic method adjustment for ties here?

Reply:
> I think the post from SO means that if ties are considered in the exact calculation, calculating the exact distribution takes longer. Our implementation of the exact distribution doesn't seem to account for ties, so my understanding is that execution time would not be affected. Would you agree?

Yes, you are right.

> I think what we should do here depends on whether we do some sort of tie correction in the asymptotic method. For instance, our `wilcoxon` does what you have here, but I see that the asymptotic method adjusts for ties (see `n_zero`). From that perspective, I can see why, in the presence of ties, the asymptotic method might be more correct than the exact method. Does the asymptotic method adjustment for ties here?

Ah, I get it now. Thanks! I think the current method does adjust for ties:

```python
2257     if repeats:   # adjust variance estimates
2258         # compute np.sum(tj * rj**2,axis=0)
2259         fac = np.sum(symrank**2, axis=0)
2260         if N % 2:  # N odd
2261             varAB = m * n * (16*N*fac - (N+1)**4) / (16.0 * N**2 * (N-1))
2262         else:  # N even
2263             varAB = m * n * (16*fac - N*(N+2)**2) / (16.0 * N * (N-1))
```

In that case, a sensible thing to do is to raise a warning when ties are present in `mode='exact'` that says something like: `The exact method doesn't correct for ties. Please use mode='auto' or mode='approx' instead if you want to account for ties.` and use the exact method only. Or we could always fall back to the approximate method (like R) after giving a warning (like `exact method doesn't account for ties. Using approximate method instead.`). IMO, the latter option is better but the former also doesn't sound bad. What do you think would be better to do here?

Reply:
I think it's preferable to stick with the exact method when the user requests the exact method.

The question I see is whether to emit a warning or not. 

But maybe it's worth a message to the mailing list. Would you be willing to send one?

Reply:
> But maybe it's worth a message to the mailing list. Would you be willing to send one?

Sure, I will write an email sometime today.

Reply:
@mdhaber I have made a mail here: https://mail.python.org/pipermail/scipy-dev/2021-March/024612.html

Reply:
Thanks!

Reply:
> I think it's preferable to stick with the exact method when the user requests the exact method.
> ...
> neither "exact" nor "asymyptotic" would check for anything (or emit any warnings); they would just respect the user's decision.

Something doesn't seem right here. Currently it's not possible to use the exact method if there are ties. It's a little unclear if that's because it's impossible to do that calculation, or because the tie handling is not implemented. Either way, just closing your eyes and doing the exact calculation while ignoring the presence of ties will lead to a wrong result, right?

Reply:
I can't answer for `ansari` in particular, but I assume it's the same as some other tests:

I think it depends on what one considers to be "wrong". For some tests like this, it's a question of which approximation you prefer to make - that there or no ties (exact), or that you have an infinitude of samples (asymptotic)?

Neither is exactly right when there are ties. For `auto`, we choose the least worst according to a published heuristic. I'm suggesting that it's not wrong to allow the user to manually override that choice. 

Yes, there are often ways of considering ties in the exact  test, but i don't think any of our exact tests implement them. I think they tend to be more computationally expensive.

Reply:
> I think it depends on what one considers to be "wrong". For some tests like this, it's a question of which approximation you prefer to make

I don't think that is right. You can either take the computationally cheap approximation, or you can implement the exact math - which is complex when there are ties, and I believe involves conditional probabilities. 

> Yes, there are often ways of considering ties in an exact test, but i don't think any of our exact tests implement them.

Exactly, that's the issue here. There is a way to work out a correct answer, and we don't have it. If there are many ties and a small sample size, the result can be way off. Rather than just give an implementation-defined answer, it looks to me like it's better to raise an exception.

https://stats.stackexchange.com/questions/11402/why-are-ties-so-difficult-in-nonparametric-statistics is informative - saying `ansari` is not the right tool here and letting people either be aware that they have to choose the approximate answer or that they should use something like a permutation test seems like the better option.

@josef-pkt you have a much better grasp of this than I do - does the above make sense?

Reply:
Ok. Should we also give an exception when there are ties and too few samples for the asymptotic approximation to be valid?

Reply:
I think we do that in other tests as well, and it seems like a good idea to have a lower bound on number of samples. The question is what that bound should be (lower than 55 for sure).

Reply:
> Should we also give an exception when there are ties and too few samples for the asymptotic approximation to be valid?

Yes, I guess that would be helpful.

> I think we do that in other tests as well, and it seems like a good idea to have a lower bound on number of samples. The question is what that bound should be (lower than 55 for sure).

Thanks for the discussion @rgommers and @mdhaber. The current behaviour is to throw an exception in presence of ties when `mode='exact'`. Is there a statistical lower bound to it or should we choose a reasonable arbitrary number? Any other suggesstions?

Reply:
I don't really have any suggestions in this direction. I was mostly interested in adding the `alternative` parameter, but this would make the PR a bit more complicated than that. If the scope is expanded to protecting users against inappropriate use of asymptotic approximations, this raises some difficult questions about how accurate is accurate enough for everyone, how to mitigate a potentially backward-incompatible change, etc. And I imagine there are a few other functions that would need to be investigated along these lines.

Reply:
> Probably the easier option would be to actually implement an exact test including ties. I did whip something together today using the code in [gh-13661](https://github.com/scipy/scipy/pull/13661), but a direct method like this probably wouldn't be fast enough to go up to n ~ 55, even if most of the computation were Pythranized as offered by @jb-leger. I think there would need to be some research into a recursive or other efficient method.

I will look into some papers and books to find a recursive/efficient implementation and let you know if get something. Also, is Pythran preferred over Cython? If there are performance concerns, should we go for Cython or Pythran?

Reply:
You can use either, depending on your preference. If you don't know either tool, Pythran is probably easier for this problem. We won't want custom parallelism here, which was the reason to go with Cython for `stats.qmc.discrepancy`.

Reply:
> You can use either, depending on your preference. If you don't know either tool, Pythran is probably easier for this problem

Ah, I see. Thanks!



> We won't want custom parallelism here, which was the reason to go with Cython for `stats.qmc.discrepancy`.

Got it.

Reply:
@tirthasheshpatel What do you think of separating the addition of the `alternative` parameter from the addition of `mode` and making a separate PR from the latter? I think it would be good to get the `alternative` parameter done now, while the experience of implementing it in other tests is fresh. Improvements to the exact test will be mostly unrelated, and they will take more time to implement and review, so a separate PR would be good.

(Personally, I don't think there is a problem in merging this more or less as-is. This function - like many others in `scipy.stats` - has been providing asymptotic p-values that may not be appropriate for small samples for almost 20 years. R's equivalent does it, too. Addition of a `'mode'` parameter is a fine time to assess whether that's OK or not, but it's not the cause of the problem. This PR just gives users the chance to pick their poison. 

As for whether the `exact` parameter with our current implementation - ignoring ties - should raise an exception or a warning or nothing at all, there are valid points for all. I think the most important thing would be to clearly document the behavior. 

I'd note that the asymptotic p-values tend to be smaller than the exact p-values, and the exact p-values ignoring ties tend to be larger than the exact p-values considering ties, so use of our current exact method may very well be preferable to our asymptotic method for small samples that include ties. It is not _accurate_ in case of ties, but the p-values are "conservative" approximations, in that they are less likely to be considered statistically significant.

But it is fine with me if we want to wait to add `mode` until we add an exact method that considers ties. That is definitely the best option.)

Reply:
> What do you think of separating the addition of the `alternative` parameter from the addition of `mode` and making a separate PR from the latter?

Agreed.

> Improvements to the exact test will be mostly unrelated, and they will take more time to implement and review, so a separate PR would be good.

Yeah, improving the exact method seems too much for this PR. I tried to find something for handling ties in the exact method for `ansari` in an efficient way but couldn't find anything other than a couple of papers to which I don't have access. Hopefully, this can be taken up in a future PR!

OK, so I will make some final tweaks to this PR:
- removing the mode parameter
- removing relevant tests
- fixing up the null/alternate hypothesis in the documentation
- reverting my `cind` change.
- using `_normtest_finish`

Anything I missed?

Reply:
@mdhaber I have made all the changes. How does everything look now?